docs/proposals/blocks-storage-sharding.md

```diff
@@ -0,0 +1,84 @@
+---
+title: "Blocks storage sharding"
+linkTitle: "Blocks storage sharding"
+weight: 1
+slug: blocks-storage-sharding
+---
+
+- Author: [Marco Pracucci](https://github.com/pracucci)
+- Date: March 2020
+- Status: proposal
+
+## Problem
+
+In Cortex, when using the experimental blocks storage, each querier internally runs the Thanos [`BucketStore`](https://github.com/thanos-io/thanos/blob/master/pkg/store/bucket.go). This means that each querier has a full view over all blocks in the long-term storage and all blocks index headers are loaded in each querier memory. The querier memory usage linearly increase with number and size of all blocks in the storage, imposing a scalability limit to the blocks storage.
```

Comment:
We also observe a quite high memory utilization in our cluster. We have a cluster with 6M active series, 1d blocks, 45 days data so far (13 months is the retention) and the Go heap inuse by `newFileBinaryReader` is 3.5GB. Projecting this to 13 months retention will be 30GB (and 6M active series is quite small).

The 1d block range period is definitely a factor here, but it's a choice made because it couples well with the 1d interval splitting by the query-frontend (once we'll have the query sharding which is the 2nd purpose of this work).

Reply:
Does the 1/32 symbols, 1/32 posting offset knowledge holds here? 6 millions in each block?

`3758096384` bytes for 6M series in 1d blocks for 45d: `83513252.977777778` per block, `13.918875496` per single series. I think this is fair 1/32 symbols 1/32 posting offset is fair assumption assuming single symbol is like usually ~30B + `uint32` and single posting offset for  binary header is

```
map[string]*postingValueOffsets

type postingValueOffsets struct {
	offsets       []postingOffset
	lastValOffset int64
}

type postingOffset struct {
	// label value.
	value string
	// offset of this entry in posting offset table in index-header file.
	tableOff int
}
```
... so like 100B or something. 

Still 3.5GB is not much for such amount of series and almost 2month of data :man_shrugging:

Reply:
Heh, but we need to scale to 120Mil active series with 13months of retention ;) If not more.